flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorPartitionLifecycleTest.java

```diff
@@ -0,0 +1,454 @@
+			.setJobManagerTable(jobManagerTable)
+			.setShuffleEnvironment(jobAwareShuffleEnvironment)
+			.setTaskSlotTable(createTaskSlotTable())
+			.build();
+
+		final TestingTaskExecutor taskManager = createTestingTaskExecutor(taskManagerServices);
+
+		try {
+			taskManager.start();
+			taskManager.waitUntilStarted();
+
+			assertTrue(jobManagerTable.contains(jobId));
+
+			taskManager.releasePartitions(jobId, Collections.singletonList(new ResultPartitionID()));
+			// connection should be kept alive since the environment still says we have local resources
+			assertTrue(jobManagerTable.contains(jobId));
+
+			hasPartitionsOccupyingLocalResources.set(false);
+
+			// the TM should check whether partitions are still stored, and afterwards terminate the connection
+			taskManager.releasePartitions(jobId, Collections.singletonList(new ResultPartitionID()));
+			assertFalse(jobManagerTable.contains(jobId));
+		} finally {
+			RpcUtils.terminateRpcEndpoint(taskManager, timeout);
+		}
+	}
+
+	@Test
+	public void testConnectionTerminationAfterInternalRelease() throws Exception {
+		final JobMasterGateway jobMasterGateway = new TestingJobMasterGatewayBuilder().build();
+
+		final JobManagerConnection jobManagerConnection = TaskSubmissionTestEnvironment.createJobManagerConnection(
+			jobId, jobMasterGateway, rpc, new NoOpTaskManagerActions(), timeout);
+
+		final JobManagerTable jobManagerTable = new JobManagerTable();
+		jobManagerTable.put(jobId, jobManagerConnection);
+
+		final AtomicBoolean hasPartitionsOccupyingLocalResources = new AtomicBoolean(true);
+		final TestJobAwareShuffleEnvironment jobAwareShuffleEnvironment = new TestJobAwareShuffleEnvironment(jobId -> hasPartitionsOccupyingLocalResources.get());
+
+		final TaskManagerServices taskManagerServices = new TaskManagerServicesBuilder()
+			.setJobManagerTable(jobManagerTable)
+			.setShuffleEnvironment(jobAwareShuffleEnvironment)
+			.setTaskSlotTable(createTaskSlotTable())
+			.build();
+
+		final TestingTaskExecutor taskManager = createTestingTaskExecutor(taskManagerServices);
+
+		try {
+			taskManager.start();
+			taskManager.waitUntilStarted();
+
+			assertTrue(jobManagerTable.contains(jobId));
+			assertNotNull(jobAwareShuffleEnvironment.listener);
+
+			jobAwareShuffleEnvironment.listener.accept(jobId);
+			// connection should be kept alive since the environment still says we have local resources
+			assertTrue(jobManagerTable.contains(jobId));
+
+			// simulates the release of internal partitions
+			hasPartitionsOccupyingLocalResources.set(false);
+
+			// the TM should terminate the connection since no more partitions are stored.
+			jobAwareShuffleEnvironment.listener.accept(jobId);
+
+			// connection termination happens asynchronously
+			final Deadline deadline = Deadline.fromNow(Duration.ofSeconds(5));
+			while (jobManagerTable.contains(jobId) && deadline.hasTimeLeft()){
+				Thread.sleep(50);
+			}
+			assertFalse(jobManagerTable.contains(jobId));
```

Comment:
Better to wait on a future which is completed by the `TestingJobMasterGateway`

Reply:
`taskManagerDisconnectFuture` is not used